src/software/new_geom/cubic_bezier_spline2d.h

```diff
@@ -0,0 +1,124 @@
+     * @param intermediate_knots The intermediate points between `start_point` and
+     *                                    `end_point` that this spline will interpolate
+     */
+    explicit CubicBezierSpline2d(const Point& start_point, const Vector& start_vector,
+                                 const Point& end_point, const Vector& end_vector,
+                                 const std::vector<Point>& intermediate_knots);
+
+    const Point getValueAt(double t) const override;
+    const std::vector<Point> getKnots() const override;
+    size_t getNumKnots() const override;
+    std::vector<double> getKnotParametrizationValues() const override;
+    const Point getStartPoint() const override;
+    const Point getEndPoint() const override;
+    const std::vector<SplineSegment2d> getSplineSegments() const override;
+
+    /**
+     * Get the control points that are the implicit representation of the bezier
+     * curves that form this spline.
+     *
+     * @return The control points that represent the the bezier curves that form this
+     *         spline.
+     */
+    const std::vector<Point>& getControlPoints() const;
+
+    /**
+     * Get the number of segments on this spline.
+     *
+     * @return The number of segments on this spline.
+     */
+    size_t getNumSegments() const;
+
+
+   private:
+    /**
+     * Get the segment of this spline at the given index
+     *
+     * @param index The index of the segment on this spline. Must be in
+     *              [0, getNumSegments()-1]
+     *
+     * @return The segment of this spline at the given index
+     */
+    BezierCurve2d getSegmentAtIndex(size_t index) const;
+
+    /**
+     * Compute the control points that represent the bezier curves that satisfy the
+     * constraints of this spline
+     *
+     * @param start_point The first point on the spline
+     * @param start_vector The tangent vector at the start of the spline
+     * @param end_point The last point on the spline
+     * @param end_vector The tangent vector at the end of the spline.
+     *                   NOTE: this points "back towards" the start of the spline, instead
+     *                         of "off the end" of the spline. ie. this vector is
+     *                         "symmetric" with `start_vector`
+     * @param intermediate_knots The intermediate points between `start_point` and
+     *                                    `end_point` that this spline will interpolate
+     * @return A vector of control points that are the implicit representation of the
+     *         bezier curves that make up the spline represented by the given args. ie.
+     *         0th-3rd (inclusive) points are the control points of the first bezier
+     *         curve, 3rd-6th (inclusive) points are the control points of the second
+     *         bezier curve, etc.
+     */
+    std::vector<Point> computeControlPoints(const Point& start_point,
```

Comment:
Any reason why we separate the start and end points into their own parameters? 

Seems like many of the comments kinda of indicate that these points are technically just the 'first' and 'last' elements of the knots vector.

Reply:
Yeah, they effectively are. We force them to be seperate in the constructor so that it's not possible to construct an invalid spline, and I just carried them through to this function because it's easier. There is _some_ special handling in this function for first/last points so I don't think it's totally ridiculous.